src/EnergyPlus/SurfaceGeometry.cc

```diff
@@ -3011,15 +2912,67 @@ namespace SurfaceGeometry {
+        // Set Zone and Space Surface First/Last Pointers
+        // Space surface lists have been built earlier in createSpaceSurfaceLists
+        for (int ZoneNum = 1; ZoneNum <= state.dataGlobal->NumOfZones; ++ZoneNum) {
+            for (int spaceNum : state.dataHeatBal->Zone(ZoneNum).spaceIndexes) {
+                auto &thisSpace = state.dataHeatBal->space(spaceNum);
+                for (int SurfNum : thisSpace.surfaces) {
+                    if (thisSpace.AllSurfaceFirst == 0) {
+                        thisSpace.AllSurfaceFirst = SurfNum;
+                    }
+                    thisSpace.AllSurfaceLast = SurfNum;
+
+                    if (state.dataSurface->Surface(SurfNum).IsAirBoundarySurf) {
+                        state.dataSurface->Surface(SurfNum).HeatTransSurf = false;
+                        continue;
+                    }
+                    // Non window surfaces are grouped next within each space
+                    if (thisSpace.HTSurfaceFirst == 0) {
+                        thisSpace.HTSurfaceFirst = SurfNum;
+                        thisSpace.OpaqOrIntMassSurfaceFirst = SurfNum;
+                        thisSpace.OpaqOrWinSurfaceFirst = SurfNum;
+                    }
+                    thisSpace.HTSurfaceLast = SurfNum;
+
+                    // Window surfaces are grouped next within each space
+                    if ((state.dataSurface->Surface(SurfNum).Class == DataSurfaces::SurfaceClass::Window) ||
+                        (state.dataSurface->Surface(SurfNum).Class == DataSurfaces::SurfaceClass::GlassDoor) ||
+                        (state.dataSurface->Surface(SurfNum).Class == DataSurfaces::SurfaceClass::TDD_Diffuser)) {
+                        if (thisSpace.WindowSurfaceFirst == 0) {
+                            thisSpace.WindowSurfaceFirst = SurfNum;
+                        }
+                        thisSpace.WindowSurfaceLast = SurfNum;
+                    } else if (state.dataSurface->Surface(SurfNum).Class != DataSurfaces::SurfaceClass::TDD_Dome) {
+                        thisSpace.OpaqOrIntMassSurfaceLast = SurfNum;
+                    }
+
+                    // TDDDome surfaces are grouped last within each space
+                    if (state.dataSurface->Surface(SurfNum).Class == DataSurfaces::SurfaceClass::TDD_Dome) {
+                        if (thisSpace.TDDDomeFirst == 0) {
+                            thisSpace.TDDDomeFirst = SurfNum;
+                        }
+                        thisSpace.TDDDomeLast = SurfNum;
+                    } else {
+                        thisSpace.OpaqOrWinSurfaceLast = SurfNum;
+                    }
+                }
+                state.dataHeatBal->Zone(ZoneNum).AllSurfaceLast = thisSpace.AllSurfaceLast;
+            }
+            int firstSpaceNum = state.dataHeatBal->Zone(ZoneNum).spaceIndexes(1);
+            state.dataHeatBal->Zone(ZoneNum).AllSurfaceFirst = state.dataHeatBal->space(firstSpaceNum).AllSurfaceFirst;
+        }
+    }
```

Comment:
It's a little difficult to grasp the changes since the code is moved entirely, but I understand that this is a positive move.  And your unit tests are fleshed out heavily, so that builds confidence.